File: examples/flow_tracker/python/flow_tracker.py
# SPDX-FileCopyrightText: Copyright (c) 2022-2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from holoscan.conditions import CountCondition
from holoscan.core import Application, Operator, OperatorSpec, Tracker

# define a custom class to represent data used in the app


class ValueData:
    """Example of a custom Python class"""

    def __init__(self, value):
        self.data = value

    def __repr__(self):
        return f"ValueData({self.data})"

    def __eq__(self, other):
        return self.data == other.data

    def __hash__(self):
        return hash(self.data)


# define custom Operators for use in the demo


class PingTxOp(Operator):
    """Simple transmitter operator.

    This operator has:
        output: "out"

    On each tick, it transmits a `ValueData` object
    """

    def __init__(self, fragment, *args, **kwargs):
        self.index = 1
        # Need to call the base class constructor last
        super().__init__(fragment, *args, **kwargs)

    def setup(self, spec: OperatorSpec):
        spec.output("out")

    def compute(self, op_input, op_output, context):
        value = ValueData(self.index)
        self.index += 1
        op_output.emit(value, "out")


class PingMxOp(Operator):
    """Example of an operator modifying data.

    This operator has:
        input:  "in"
        output: "out"

    The data from each input is multiplied by a user-defined value.
    """

    def __init__(self, fragment, *args, **kwargs):
        self.count = 1

        # Need to call the base class constructor last
        super().__init__(fragment, *args, **kwargs)

    def setup(self, spec: OperatorSpec):
        spec.input("in")
        spec.output("out")
        spec.param("multiplier", 2)

    def compute(self, op_input, op_output, context):
        value = op_input.receive("in")
        print(f"Middle message received (count: {self.count})")
        self.count += 1

        print(f"Middle message value: {value.data}")

        # Multiply the values by the multiplier parameter
        value.data *= self.multiplier

        op_output.emit(value, "out")


class PingRxOp(Operator):
    """Simple receiver operator.

    This operator has:
        input: "receivers"

    This is an example of a native operator that can dynamically have any
    number of inputs connected to is "receivers" port.
    """

    def __init__(self, fragment, *args, **kwargs):
        self.count = 1
        # Need to call the base class constructor last
        super().__init__(fragment, *args, **kwargs)

    def setup(self, spec: OperatorSpec):
        spec.param("receivers", kind="receivers")

    def compute(self, op_input, op_output, context):
        values = op_input.receive("receivers")
        print(f"Rx message received (count: {self.count}, size: {len(values)})")
        self.count += 1

        print(values)
        print(f"Rx message value0: {values[0].data}")


# Now define a simple application using the operators defined above


class MyPingApp(Application):
    def compose(self):
        root1 = PingTxOp(self, CountCondition(self, 10), name="root1")
        root2 = PingTxOp(self, CountCondition(self, 10), name="root2")
        middle1 = PingMxOp(self, name="middle1", multiplier=2)
        middle2 = PingMxOp(self, name="middle2", multiplier=3)
        leaf1 = PingRxOp(self, name="leaf1")
        leaf2 = PingRxOp(self, name="leaf2")

        # Define the workflow
        self.add_flow(root1, middle1)
        self.add_flow(middle1, leaf1, {("out", "receivers")})
        self.add_flow(middle1, leaf2, {("out", "receivers")})

        self.add_flow(root2, middle2)
        self.add_flow(middle2, leaf2, {("out", "receivers")})


if __name__ == "__main__":
    app = MyPingApp()
    with Tracker(
        app, filename="logger.log", num_start_messages_to_skip=2, num_last_messages_to_discard=3
    ) as tracker:
        app.run()
        tracker.print()

File: examples/flow_tracker/cpp/flow_tracker.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <memory>
#include <vector>

#include "holoscan/holoscan.hpp"

class ValueData {
 public:
  ValueData() = default;
  explicit ValueData(int value) : data_(value) {
    HOLOSCAN_LOG_TRACE("ValueData::ValueData(): {}", data_);
  }
  ~ValueData() { HOLOSCAN_LOG_TRACE("ValueData::~ValueData(): {}", data_); }

  // Use default copy constructor
  ValueData(const ValueData&) = default;

  // Use default move constructor
  ValueData(ValueData&&) noexcept = default;

  // Use default copy assignment operator
  ValueData& operator=(const ValueData&) = default;

  // Use default move assignment operator
  ValueData& operator=(ValueData&&) noexcept = default;

  void data(int value) { data_ = value; }

  [[nodiscard]] int data() const { return data_; }

 private:
  int data_;
};

namespace holoscan::ops {

class PingTxOp : public Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(PingTxOp)

  PingTxOp() = default;

  void setup(OperatorSpec& spec) override { spec.output<ValueData>("out"); }

  void compute([[maybe_unused]] InputContext& op_input, OutputContext& op_output,
               [[maybe_unused]] ExecutionContext& context) override {
    auto value1 = ValueData(index_++);
    op_output.emit(value1, "out");
  };

 private:
  int index_ = 1;
};

class PingMxOp : public Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(PingMxOp)

  PingMxOp() = default;

  void setup(OperatorSpec& spec) override {
    spec.input<ValueData>("in");
    spec.output<ValueData>("out");
    spec.param(multiplier_, "multiplier", "Multiplier", "Multiply the input by this value", 2);
  }

  void compute(InputContext& op_input, OutputContext& op_output,
               [[maybe_unused]] ExecutionContext& context) override {
    auto value = op_input.receive<ValueData>("in").value();

    HOLOSCAN_LOG_INFO("Middle message received (count: {})", count_++);

    HOLOSCAN_LOG_INFO("Middle message value1: {}", value.data());

    // Multiply the values by the multiplier parameter
    value.data(value.data() * multiplier_);

    op_output.emit(value, "out");
  };

 private:
  int count_ = 1;
  Parameter<int> multiplier_;
};

class PingRxOp : public Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(PingRxOp)

  PingRxOp() = default;

  void setup(OperatorSpec& spec) override {
    // // Since Holoscan SDK v2.3, users can define a multi-receiver input port using 'spec.input()'
    // // with 'IOSpec::kAnySize'.
    // // The old way is to use 'spec.param()' with 'Parameter<std::vector<IOSpec*>> receivers_;'.
    // spec.param(receivers_, "receivers", "Input Receivers", "List of input receivers.", {});
    spec.input<std::vector<ValueData>>("receivers", IOSpec::kAnySize);
  }

  void compute(InputContext& op_input, [[maybe_unused]] OutputContext& op_output,
               [[maybe_unused]] ExecutionContext& context) override {
    auto value_vector = op_input.receive<std::vector<ValueData>>("receivers").value();

    HOLOSCAN_LOG_INFO("Rx message received (count: {}, size: {})", count_++, value_vector.size());

    for (int i = 0; i < value_vector.size(); i++) {
      HOLOSCAN_LOG_INFO("Rx message value{}: {}", i, value_vector[i].data());
    }
  };

 private:
  // // Since Holoscan SDK v2.3, the following line is no longer needed.
  // Parameter<std::vector<IOSpec*>> receivers_;
  int count_ = 1;
};

}  // namespace holoscan::ops

class MyPingApp : public holoscan::Application {
 public:
  void compose() override {
    using namespace holoscan;

    // Define the tx, mx, rx operators, allowing the tx operator to execute 10 times
    auto root1 = make_operator<ops::PingTxOp>("root1", make_condition<CountCondition>(10));
    auto root2 = make_operator<ops::PingTxOp>("root2", make_condition<CountCondition>(10));
    auto middle1 = make_operator<ops::PingMxOp>("middle1", Arg("multiplier", 2));
    auto middle2 = make_operator<ops::PingMxOp>("middle2", Arg("multiplier", 3));
    auto leaf1 = make_operator<ops::PingRxOp>("leaf1");
    auto leaf2 = make_operator<ops::PingRxOp>("leaf2");

    // Create the workflow graph
    add_flow(root1, middle1);
    add_flow(middle1, leaf1, {{"out", "receivers"}});
    add_flow(middle1, leaf2, {{"out", "receivers"}});

    add_flow(root2, middle2);
    add_flow(middle2, leaf2, {{"out", "receivers"}});
  }
};

int main([[maybe_unused]] int argc, [[maybe_unused]] char** argv) {
  auto app = holoscan::make_application<MyPingApp>();
  // Skip 2 messages at the start and 3 messages at the end
  auto& tracker = app->track(2, 3, 0);
  tracker.enable_logging();
  app->run();

  // Print all the metrics to the standard output
  tracker.print();

  // The code below shows how individual metrics can be accessed via the API
  // Get all the path strings
  auto path_strings = tracker.get_path_strings();
  // Print maximum end-to-end latency for every path
  HOLOSCAN_LOG_INFO("Maximum Latencies:");
  for (auto& path_string : path_strings) {
    HOLOSCAN_LOG_INFO("Path: {} -- {} ms",
                      path_string,
                      tracker.get_metric(path_string, holoscan::DataFlowMetric::kMaxE2ELatency));
  }

  auto root_message_number = tracker.get_metric(holoscan::DataFlowMetric::kNumSrcMessages);

  for (auto& root_message_number_pair : root_message_number) {
    HOLOSCAN_LOG_INFO("Root-transmitter: {}, number of messages: {}",
                      root_message_number_pair.first,
                      root_message_number_pair.second);
  }
  return 0;
}
